File: python/looker/rtl/auth_token.py
from typing import Optional
import datetime

from looker.sdk import models as ml


class AuthToken():
    """Used to instantiate or check expiry of an AccessToken object"""

    def __init__(self, token: Optional[ml.AccessToken] = None):
        self.access_token: str = ''
        self.token_type: str = ''
        self.expires_in: int = 0
        self.expires_at = datetime.datetime.now()
        if token is None:
            token = ml.AccessToken()
        self.set_token(token)

    def set_token(self, token: ml.AccessToken):
        """Assign the token and set its expiration."""
        self.access_token = token.access_token or ''
        self.token_type = token.token_type or ''
        self.expires_in = token.expires_in or 0

        exp = datetime.datetime.now()

        if (token.access_token and token.expires_in):
            exp = exp + datetime.timedelta(seconds=token.expires_in)
        else:
            # set to expire 10 seconds ago
            exp = exp + datetime.timedelta(seconds=-10)
        self.expires_at = exp

    @property
    def is_active(self) -> bool:
        """True if authentication token has not timed out"""
        if not self.expires_at:
            return False
        return self.expires_at > datetime.datetime.now()

File: csharp/rtl/AuthToken.cs
using System;
using System.Collections;
using System.Globalization;
using System.Security.Permissions;

#nullable enable

namespace Looker.RTL
{
    // TODO: replace IAccessToken with generated AccessToken interface
    // TODO: figure out nullable ints
    public interface IAccessToken
    {
        string? access_token { get; set; }
        string? token_type { get; set; }
        int expires_in { get; set; }
        string? refresh_token { get; set; }
    }

    /// <summary>
    /// Concrete implementation of IAccessToken
    /// </summary>
    public class AccessToken : IAccessToken
    {
        public string? access_token { get; set; }
        public string? token_type { get; set; }
        public int expires_in { get; set; }
        public string? refresh_token { get; set; }
    }

    /// <summary>
    /// Used to instantiate or check expiry of an AccessToken struct
    /// </summary>
    public class AuthToken: IAccessToken
    {
        private static long lagTime = 10;
        public string? access_token { get; set; }
        public string? token_type { get; set; }

        public int expires_in { get; set; }

        public string? refresh_token { get; set; }
        // Give a 10 second expiration window to account for latency
        public DateTime ExpiresAt { get; set; } = DateTime.Now.AddSeconds(-lagTime);

        public AuthToken() { }

        public AuthToken(IAccessToken token)
        {
            SetToken(token);
        }

        /// <summary>
        /// Assign token and set its expiration
        /// </summary>
        /// <param name="token"></param>
        public void SetToken(IAccessToken token)
        {
            if (token.token_type.IsFull())
            {
                token_type = token.token_type;
            }

            if (token.refresh_token.IsFull())
            {
                refresh_token = token.refresh_token;
            }

            expires_in = token.expires_in;
            access_token = token.access_token;
            ExpiresAt = DateTime.Now.AddSeconds(expires_in > 0 ? expires_in - lagTime : -lagTime);
        }

        /// <summary>
        /// True if authentication token has not timed out
        /// </summary>
        /// <returns></returns>
        public bool IsActive()
        {
            if (access_token.IsNullOrEmpty())
            {
                return false;
            }
            return ExpiresAt > DateTime.Now;
        }

        public AuthToken Reset()
        {
            access_token = null;
            expires_in = 0;
            ExpiresAt = DateTime.Now.AddSeconds(-lagTime);
            return this;
        }
    }
}
